src/main/kotlin/org/elm/lang/core/psi/PsiElement.kt

```diff
@@ -72,12 +73,106 @@ fun <T : PsiElement> PsiElement.parentOfType(vararg classes: KClass<out T>): T?
 inline fun <reified T : PsiElement> PsiElement.contextOfType(strict: Boolean = true): T? =
         PsiTreeUtil.getContextOfType(this, T::class.java, strict)
 
-inline fun <reified T : PsiElement> PsiElement.childOfType(strict: Boolean = true): T? =
-        PsiTreeUtil.findChildOfType(this, T::class.java, strict)
+inline fun <reified T : PsiElement> PsiElement.descendantOfType(): T? =
+        PsiTreeUtil.findChildrenOfType(this, T::class.java).find { it is T }
 
 inline fun <reified T : PsiElement> PsiElement.descendantsOfType(): Collection<T> =
         PsiTreeUtil.findChildrenOfType(this, T::class.java)
 
+/** Returns direct children of the specified type */
+inline fun <reified T : PsiElement> PsiElement.directChildrenOfType(): List<T> =
+        PsiTreeUtil.getChildrenOfTypeAsList(this, T::class.java)
+
+/** Returns direct children stubs (or AST, if necessary) of the specified type */
+inline fun <reified T : PsiElement> PsiElement.stubDirectChildrenOfType(): List<T> {
+    return if (this is PsiFileImpl) {
+        stub?.childrenStubs?.mapNotNull { it.psi as? T } ?: return directChildrenOfType()
+    } else {
+        PsiTreeUtil.getStubChildrenOfTypeAsList(this, T::class.java)
+    }
+}
+
+inline fun <reified T : PsiElement> PsiElement.stubDescendantsOfTypeStrict(): Collection<T> =
```

Comment:
Maybe use a `strict` parameter rather than separate functions for the strict case? It's fine either way.

Reply:
I copied this straight from intellij-rust. I don't have a preference either, at least not yet.